### PR TITLE
Map wires #2: Add top-level `qml.map_wires` function

### PR DESCRIPTION
**Pull request by AlbertMitjans**
**NOTE:** Review after the following PR is merged into master:
- https://github.com/PennyLaneAI/pennylane/pull/3143

Added the `qml.map_wires` function, that changes the wires of the given operator, `QNode`, `QuantumTape` or quantum function according to the given wire map.

  Using `qml.map_wires` with an operator:

  ```pycon
  >>> op = qml.RX(0.54, wires=0) + qml.PauliX(1) + (qml.PauliZ(2) @ qml.RY(1.23, wires=3))
  >>> op
  (RX(0.54, wires=[0]) + PauliX(wires=[1])) + (PauliZ(wires=[2]) @ RY(1.23, wires=[3]))
  >>> wire_map = {0: 3, 1: 2, 2: 1, 3: 0}
  >>> qml.map_wires(op, wire_map)
  (RX(0.54, wires=[3]) + PauliX(wires=[2])) + (PauliZ(wires=[1]) @ RY(1.23, wires=[0]))
  ```

  Using `qml.map_wires` with a `QNode`:

  ```pycon
  >>> dev = qml.device("default.qubit", wires=4)
  >>> @qml.qnode(dev)
      def circuit():
          qml.RX(0.54, wires=0) @ qml.PauliX(1) @ qml.PauliZ(2) @ qml.RY(1.23, wires=3)
          return qml.probs(wires=0)
  >>> mapped_circuit = qml.map_wires(circuit, wire_map)
  >>> mapped_circuit()
  tensor([0.92885434, 0.07114566], requires_grad=True)
  >>> list(mapped_circuit.tape)
  [((RX(0.54, wires=[3]) @ PauliX(wires=[2])) @ PauliZ(wires=[1])) @ RY(1.23, wires=[0]),
  probs(wires=[3])]
  ```

### CODE REVIEW

**Comment by AlbertMitjans** [comment]
[sc-27112]

File: doc/releases/changelog-dev.md
Comment (albi3ro) [inline doc/releases/changelog-dev.md:72]:
```suggestion
  >>> @qml.qnode(dev)
  ... def circuit():
  ...     qml.RX(0.54, wires=0) @ qml.PauliX(1) @ qml.PauliZ(2) @ qml.RY(1.23, wires=3)
  ...     return qml.probs(wires=0)
```

Comment (albi3ro) [reply doc/releases/changelog-dev.md:72]: Actually, for these examples, it's probably best to:
* Use plain ops, not operator arithmetic
* Show the circuit using `qml.draw`, not exposing the `tape`
* Use an obvious wire map, like `{0:10, 1:11, 2:12, 3:13}` or `{0:"a", 1:"b", 2:"c", 3:"d"}`

File: pennylane/ops/functions/map_wires.py
Comment (albi3ro) [inline pennylane/ops/functions/map_wires.py:32]: Good to document `queue` and `replace`.

File: doc/releases/changelog-dev.md
Comment (AlbertMitjans) [reply doc/releases/changelog-dev.md]: > - Use plain ops, not operator arithmetic
Done
> - Show the circuit using qml.draw, not exposing the tape
If we do this the wire values are not shown:
```pycon
>>> print(qml.draw(circuit)())
0: ──RX(0.54)─┤  Probs
1: ──X────────┤
2: ──Z────────┤
3: ──RY(1.23)─┤
```
> - Use an obvious wire map, like {0:10, 1:11, 2:12, 3:13} or {0:"a", 1:"b", 2:"c", 3:"d"}
Using {0:10, 1:11, 2:12, 3:13} ✅

File: pennylane/ops/functions/map_wires.py
Comment (AlbertMitjans) [reply pennylane/ops/functions/map_wires.py:32]: Done!

File: pennylane/ops/functions/map_wires.py
Comment (albi3ro) [inline pennylane/ops/functions/map_wires.py:87]:
```suggestion
    if isinstance(input, QuantumScript):
        new_qscript = QuantumScript(
            (op.map_wires(wire_map) for op in input._ops),
            (m.map_wires(wire_map) for m in input._measurements),
            (op.map_wires(wire_map) for op in input._prep)
        )
```

File: doc/releases/changelog-dev.md
Comment (albi3ro) [reply doc/releases/changelog-dev.md]: `circuit` was the unmapped qnode:

```pycon
>>> print(qml.draw(mapped_circuit)())
10: ──RX(0.54)─┤  Probs
11: ──X────────┤       
12: ──Z────────┤       
13: ──RY(1.23)─┤       
```

Had me concerned for a moment.

File: pennylane/ops/functions/map_wires.py
Comment (albi3ro) [inline pennylane/ops/functions/map_wires.py:49]: Since we aren't supporting decorator syntax for the time being, we should add a note or warning that `@qml.map_wires` *won't* work.

Comment (josh146) [reply pennylane/ops/functions/map_wires.py:49]: @albi3ro we could always use the approach JAX uses in their documentation. E.g.,

```python
@functools.partial(qml.map_wires, wire_map)
@qnode
def func(x):
   ...
```

But not too important I agree

File: tests/ops/functions/test_map_wires.py
Comment (Jaybsoni) [inline tests/ops/functions/test_map_wires.py:24]: We can probably make this private

File: tests/ops/functions/test_map_wires.py
Comment (Jaybsoni) [inline tests/ops/functions/test_map_wires.py:192]: Worth testing with other interfaces as well?

File: tests/ops/functions/test_map_wires.py
Comment (Jaybsoni) [inline tests/ops/functions/test_map_wires.py:161]: maybe change this test to measure and apply two different ops?

**Review by Jaybsoni — COMMENTED** [review]
Looks good to go for me 💯! Just a few small comments and failing tests to address then happy to approve

File: pennylane/tape/tape.py
Comment (albi3ro) [reply pennylane/tape/tape.py:19]: I might have messed up the merge conflict.

File: tests/ops/functions/test_map_wires.py
Comment (albi3ro) [reply tests/ops/functions/test_map_wires.py:24]: If it's in a test file, I don't think that matters too much.

File: doc/releases/changelog-dev.md
Comment (AlbertMitjans) [reply doc/releases/changelog-dev.md]: Shoot, sorry for that 😢

Comment (AlbertMitjans) [reply doc/releases/changelog-dev.md]: Changed!

File: pennylane/ops/functions/map_wires.py
Comment (AlbertMitjans) [reply pennylane/ops/functions/map_wires.py:62]: Added note mentioning the use of functools

File: pennylane/tape/tape.py
Comment (AlbertMitjans) [reply pennylane/tape/tape.py:19]: Imported

File: tests/ops/functions/test_map_wires.py
Comment (AlbertMitjans) [reply tests/ops/functions/test_map_wires.py:192]: This is testing that the circuit can be compiled using jax.

I don't think we should add more tests with other interfaces given that wires is independent of the interface

File: pennylane/ops/functions/map_wires.py
Comment (albi3ro) [inline pennylane/ops/functions/map_wires.py:50]:
```suggestion
        ``qml.map_wires`` can be used as a decorator with the help of the ``functools`` module:
```

File: pennylane/ops/functions/map_wires.py
Comment (albi3ro) [inline pennylane/ops/functions/map_wires.py:38]:
```suggestion
        input (.Operator, pennylane.QNode, .QuantumScript, or Callable): an operator, quantum node,
```

File: pennylane/ops/functions/map_wires.py
Comment (albi3ro) [inline pennylane/ops/functions/map_wires.py:39]:
```suggestion
        input (.Operator, pennylane.QNode, .QuantumScript, or Callable): an operator, quantum node,
            quantum script, or function that applies quantum operations
```

File: pennylane/ops/functions/map_wires.py
Comment (albi3ro) [inline pennylane/ops/functions/map_wires.py:46]:
```suggestion
        (.Operator, pennylane.QNode, .QuantumScript, or Callable): input with changed wires
```

**Review by albi3ro — APPROVED** [review]
Left a couple of comments about switching references from `QuantumTape` to `QuantumScript` in the documentation.

But looks great! Glad to have this 🚀

**Comment by AlbertMitjans** [comment]
@Jaybsoni I will wait for your approval before merging

**Review by Jaybsoni — APPROVED** [review]
💯

**Pull request merged**
